src/cpp/core/system/PosixChildProcess.cpp

```diff
@@ -660,14 +665,81 @@ struct AsyncChildProcess::AsyncImpl
       : calledOnStarted_(false),
         finishedStdout_(false),
         finishedStderr_(false),
-        exited_(false)
+        exited_(false),
+        checkSubProc_(boost::posix_time::not_a_date_time),
+        hasSubprocess_(true)
    {
    }
-
    bool calledOnStarted_;
    bool finishedStdout_;
    bool finishedStderr_;
    bool exited_;
+
+   // when we next check subprocess count
+   boost::posix_time::ptime checkSubProc_;
+
+   // result of last subprocess check
+   bool hasSubprocess_;
+
+   void initTimers(bool reportHasSubprocs)
+   {
+      if (exited_)
+      {
+         checkSubProc_ = boost::posix_time::not_a_date_time;
+         return;
+      }
+
+      if (reportHasSubprocs && checkSubProc_.is_not_a_date_time())
+      {
+         checkSubProc_ = now() + boost::posix_time::seconds(1);
+      }
+   }
+
+   bool checkChildCount()
+   {
+      if (checkSubProc_.is_not_a_date_time())
+         return false;
+
+      bool fCheck = now() > checkSubProc_;
+
+      if (fCheck)
+         checkSubProc_ = boost::posix_time::not_a_date_time;
+      return fCheck;
+   }
+
+   bool computeHasSubProcess(pid_t pid)
+   {
+#ifdef _WIN32
+      return true;
+#else
+      // pgrep -P ppid returns 0 if there are matches, non-zero
+      // otherwise
+      shell_utils::ShellCommand cmd("pgrep");
```

Comment:
It should be possible to have the kernel deliver notifications to us when the shell process forks or execs, so we can keep this updated in real time instead of polling. See e.g.:

http://netsplit.com/the-proc-connector-and-socket-filters

However it doesn't seem like this is very straightforward (and may wind up being somewhat non-portable and complicated) so maybe just worth filing a note to revisit. If we don't wind up with an event-driven approach we  also might be able to find a way to query for child processes using system calls rather than spawning a whole new process to count.